### PR TITLE
Use Source Code Generator for PSVersionInfo class

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

The custom source generator allows us to exclude reflection and to directly get current PowerShell version from MSBuild.

PerfView doesn't work well on current .Net Preview. A fix has just been merged in .Net Runtime repo for this. I hope we get it in .Net 6.0 Preview6 or Preview7. So today I can not share reliable result from PerfView to confirm perf win. Obviously using static string is more faster then reflection.

## PR Context

Contribute to #13540
Related #14268

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by iSazonov** [comment]
It seems test fails are not related to the PR - pwsh.exe has Preview.**8** in product version instead of Preview.**7**.

@rjmholt @TravisEz13 Could you please look the issue?

**Comment by rjmholt** [comment]
@iSazonov opened https://github.com/PowerShell/PowerShell/issues/15607

File: src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (rkeithhill) [inline src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:27]: Typo. Should be - `Roslyn analyzers`

Comment (iSazonov) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs]: Thanks! 

Fixed.

File: src/System.Management.Automation/engine/PSVersionInfo.cs
Comment (rkeithhill) [inline src/System.Management.Automation/engine/PSVersionInfo.cs:45]: Maybe tweak to `The following property ...`

File: src/System.Management.Automation/engine/PSVersionInfo.cs
Comment (rkeithhill) [inline src/System.Management.Automation/engine/PSVersionInfo.cs:44]: Use of multiline comment seems a bit out-of-place.  Consider using `//`

File: src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (iSazonov) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:15]: Hmm, if I understand correctly the class is used in MSBuild and should be public but I did not check. In any case the class is not a part of PowerShell itself (it is not included in any PowerShell assembly).

File: src/System.Management.Automation/engine/PSVersionInfo.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/PSVersionInfo.cs:45]: Fixed.

File: src/System.Management.Automation/engine/PSVersionInfo.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/PSVersionInfo.cs:44]: Using `//` raise CodeFactor issue (blank line after the comment).

**Comment by iSazonov** [comment]
@rjmholt I resolved the issue with release tag. It turned out that ci.psm1 always assigns ReleaseTag property directly to dotnet.

File: src/System.Management.Automation/engine/PSVersionInfo.cs
Comment (vexx32) [reply src/System.Management.Automation/engine/PSVersionInfo.cs:44]: I think CodeFactor recommends using `////` to comment out code specifically, which bypasses the usual warning for that.

Comment (iSazonov) [reply src/System.Management.Automation/engine/PSVersionInfo.cs:44]: @xtqqczze Have you thoughts?

**Review by TravisEz13 — DISMISSED** [review]
Pending @wg-security review

File: src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (iSazonov) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:70]: Notice, the file is written to disk and it is accessible for compliance tools.

@PaulHigin Friendly ping.

**Comment by PaulHigin** [comment]
@iSazonov  Sorry for the delay.  I am unavailable today but will look at this early next week.

File: src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (PaulHigin) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:70]: @TravisEz13 , @iSazonov 
The concern, of course, is code injection.  I don't immediately see an issue, but this needs to be discussed in the next WG-Security meeting.  However, a meeting may not occur until December, due to the holidays.  Since this is an optional change, I feel it is not high priority.

Comment (PaulHigin) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:70]: To be clear, this is the opinion of the @WG-Security group, not just me.  We looked at `const string SourceTemplate` where the above property values are concatenated into the source template as strings.  This is a supply chain source code injection opportunity, that would be difficult to detect.  Our understanding is that using a CodeDom is the preferred way to build up source code for compilation.

Comment (iSazonov) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:70]: @PaulHigin I really don't understand about CodeDom. The best comment from .Net team I saw about using CodeDom in SGs is https://github.com/dotnet/roslyn/issues/48214#issuecomment-704432433

Comment (PaulHigin) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:70]: I'm not an expert either, and I'll defer to @TravisEz13.

Comment (SteveL-MSFT) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:70]: For code injection to happen here, there would need to be a double quote, right?  Can we just reject any strings that contain a double quote?

Comment (iSazonov) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:70]: I really don't understand. We could implement this as TypeGen and ResGen which we already use. Only difference is the TypeGen and ResGen are our custom utilities but the source generator is .Net feature.
To do injection, someone would have to compromise the entire repository in order to corrupt both build scripts and the source code.

**Comment by daxian-dbw** [comment]
@iSazonov I really like you trying out the source generator!
I measured the code you removed from `PSVersionInfo.cs` by using the source generator, and it takes about `7ms` on average on my machine for private build (going through the `Contains(" Commit ")` code path, and hence slower). 

> ```
> Assembly currentAssembly = typeof(PSVersionInfo).Assembly;
> ProductVersion = currentAssembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>().InformationalVersion;
> 
> string rawGitCommitId;
> string mainVersion = ProductVersion.Substring(0, ProductVersion.IndexOf(' '));
> 
> if (ProductVersion.Contains(" Commits: "))
> {
>     rawGitCommitId = ProductVersion.Replace(" Commits: ", "-").Replace(" SHA: ", "-g");
> }
> else
> {
>     rawGitCommitId = mainVersion;
> }
> ```

The time it takes for `pwsh -noprofile -c exit` on my machine is about `390ms` on average (ranging from ~380 to ~420), and the time improvement is about `1.7%`.

Not sure if the gain worth the extra complexity though.

**Comment by iSazonov** [comment]
> Not sure if the gain worth the extra complexity though.

It's no more complicated than our ResGen or TypeGen. It's just something new to get used to. 😄 

We could replace some of our Regex with a standard .Net Regex SG to get benefits in startup scenario and other scenarios.
And we could also replace TypeGen with our SG, which of course will not speed up PowerShell itself, but will noticeably speed up the CIs and the generated cache will probably be more correct.

**Comment by daxian-dbw** [comment]
The current type initializer of PSVersionInfo takes `24-30 ms` on my dev machine, and it happens when creating the banner messgae, which is very early in the startup process. It would be good for the startup time if we can reduce the time it takes.

I will do an end-to-end measurement on the gains using source generator for `PSVersionInfo` plus the changes in https://github.com/PowerShell/PowerShell/pull/18210, and will report back the results.

**Review by daxian-dbw — COMMENTED** [review]
I integrated the `FastParsePSVersion` change from #18210 to the source generator, and also refactored the code. After the change, the static constructor of `PSVersionTable` only takes `2-3 ms` on my dev machine (release build with R2R images, using the release tag `v7.3.0` to mimic a stable release). It was `24-30 ms` without the changes of this PR:

```
## Measured by wrapping the static constructor with `StopWatch`.

C:\>E:\arena\source\PowerShell\src\powershell-win-core\bin\Release\net7.0\win7-x64\publish\pwsh.exe -noprofile
2.5478
PowerShell 7.3.0
PS C:\> exit

C:\>E:\arena\source\PowerShell\src\powershell-win-core\bin\Release\net7.0\win7-x64\publish\pwsh.exe -noprofile
2.5993
PowerShell 7.3.0
PS C:\> exit

C:\>E:\arena\source\PowerShell\src\powershell-win-core\bin\Release\net7.0\win7-x64\publish\pwsh.exe -noprofile
2.5842
PowerShell 7.3.0
PS C:\> exit
```

File: src/System.Management.Automation/System.Management.Automation.csproj
Comment (daxian-dbw) [inline src/System.Management.Automation/System.Management.Automation.csproj:62]: The only remaining problem is: the generated source code is not visible to IDEs like VSCode and Visual Studio.
The `<Compile Remove ... />` tag is needed because otherwise the compiler will report errors about duplicate declarations, as it automatically includes the im-memory source output, as well as the file generated from the last build.

In Visual Studio or VSCode, the constant members used in `PSVersionInfo.cs` cannot be resolved, because when using `<Compile Remove ... />`, the files will not be included for analyzing. I tried many ways, including attempting to delete the generated file using `<RemoveDir>` task in `System.Management.Automation.csproj` or `PSVersionInfoGenerator.csproj`, but nothing works. @iSazonov and @SeeminglyScience, really need help on this.

Comment (iSazonov) [reply src/System.Management.Automation/System.Management.Automation.csproj:62]: Eh, it's been so long that I don't remember all the details anymore. Yes, there was such a problem and this line was a workaround. I even opened the case https://github.com/OmniSharp/omnisharp-vscode/issues/4215 and there's a progression in recent months, but I haven't checked if it's already in the current public version of VSCode/OmniSharp.
I think I saw in the VS 2022 announcement that it has SGs support, but I haven't checked and can't say for sure.

I'll try to find something useful

Comment (iSazonov) [reply src/System.Management.Automation/System.Management.Automation.csproj]: I added cleanup commit:
- Turn off EmitCompilerGeneratedFiles - since latest VS Code works well we have no need to keep source of the generated file.
- Remove first blank line in generated code - cosmetic change

File: src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (iSazonov) [inline src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:110]: We have no need to do "fast" on some ns since it is _build_ time.

Also, it is big headache to partially process versions in so many places (Build/Release.psm1, csproj, SG, SMA) - any minor change in one place can break all unpredictable.

_I strongly suggest keeping all version parsing in PowerShell.Common.props file._

As result, the SG will be very simple and stable, and our build process will be easily maintainable.

Comment (daxian-dbw) [reply src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:110]: > We have no need to do "fast" on some ns since it is build time.

It's just using the same function name. _fast_ is to compare with letting `SemanticVersion` do the string parsing.

> I strongly suggest keeping all version parsing in PowerShell.Common.props file.

I disagree. The reason to handle version in .csproj is that we need to declare `ProductVersion` for all assemblies. But only `S.M.A` needs the parsed components of the version string, so keeping it in `PowerShell.Common.props` will increase build time for all assemblies, unnecessarily.

This change is essentially moving the version parsing code from `SMA` to source generator (`PSVersionInfo` itself doesn't do version parsing anymore). If we need to change how PowerShell version will be represented in future (which I think is unlikely), then it's also easier to change the parsing in C# code and also easier to verify the change.

File: src/System.Management.Automation/System.Management.Automation.csproj
Comment (daxian-dbw) [reply src/System.Management.Automation/System.Management.Automation.csproj]: It doesn't work in the following scenarios, for both VSCode and Visual Studio, which are quite common in development:
1. Run `Start-PSBuild` first, then open `System.Management.Automation` in VSCode
2. Open `System.Management.Automation` in VSCode to use an existing build, like from yesterday.

But I'm fine turning off `EmitCompilerGeneratedFiles` since it doesn't help. Also, I'm fine merging the PR without perfectly resolving this issue.

File: src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (iSazonov) [reply src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs]: > I disagree. The reason to handle version in .csproj is that we need to declare `ProductVersion` for all assemblies. But only `S.M.A` needs the parsed components of the version string, so keeping it in `PowerShell.Common.props` will increase build time for all assemblies, unnecessarily.

Handling version in .csproj probably takes a second or two. This is incomparable to the compile time which takes several minutes because of analyzers.

>If we need to change how PowerShell version will be represented in future (which I think is unlikely), then it's also easier to change the parsing in C# code and also easier to verify the change.

Exactly what is unlikely and this has already been concentrated in csproj for a long time. Now we are smearing it again.
When I opened this PR package tests were crashing because Release.psm1 was doing something incomprehensible. Now these tests are working, i.e. something was changed there again.
The release process is entirely on the MSFT team and above all you will waste your time over and over again. Although the community also releases their packages and they just benefit from having all the logic in csproj so they don't have to puzzle over the convoluted logic that is smeared out.

File: src/System.Management.Automation/System.Management.Automation.csproj
Comment (daxian-dbw) [reply src/System.Management.Automation/System.Management.Automation.csproj]: I pushed a new commit to solve this issue, not perfectly, but works.

File: src/System.Management.Automation/System.Management.Automation.csproj
Comment (daxian-dbw) [inline src/System.Management.Automation/System.Management.Automation.csproj:13]: Generate the file, so VSCode and Visual Studio can always resolve the generated members.
Those files will be removed when building again, by `RemoveDir` in `PowerShell.Common.props`.

File: PowerShell.Common.props
Comment (daxian-dbw) [inline PowerShell.Common.props:98]: The path `gen\SourceGenerated` is duplicated in here and `System.Management.Automation.csproj`. I tried creating a property holding the path in the `PropertyGroup` right above, and use it in `RemoveDir` and `CompilerGeneratedFilesOutputPath`. However, it works in `RemoveDir`, but not in `CompilerGeneratedFilesOutputPath`, and I cannot figure out why. I'd love if someone can make it work.

File: src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs]: > Handling version in .csproj probably takes a second or two. This is incomparable to the compile time which takes several minutes because of analyzers.

1. It's not only about a few more seconds, it's more about not doing things that are not needed. Again, all other assemblies other than `S.M.A` don't need the version to be parsed.
2. Parsing the version in C# code is more maintainable and easier to change if really needed.

> The release process is entirely on the MSFT team and above all you will waste your time over and over again

I don't know about other maintainers' experience regarding this, but I really cannot remember what the last time was I saw a build failed due to parsing the version string in C# code.

**Comment by daxian-dbw** [comment]
@TravisEz13 and @PaulHigin, can you please review the changes again and re-evaluate the "code injection" concern?

Here is the doc (and the sample code) about the source generator: https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview

The arguments used to format the source template in [PSVersionInfoGenerator.cs](https://github.com/PowerShell/PowerShell/pull/15603/files#diff-4757e7bd4af91fb154f76a6ecbb38da1ca557e1994cebfb674e355a64f6169e1) are from the build properties defined in `PowerShell.Common.props`.

File: src/System.Management.Automation/System.Management.Automation.csproj
Comment (iSazonov) [reply src/System.Management.Automation/System.Management.Automation.csproj:13]: I pushed a commit a day before (but you reverted now) and removed this old workaround because all works well in VS Code. We have no need to emit generated files.
Perhaps you need re-configure your local environment.
I use VS Code 1.72 and C# extension 1.25
image

Comment (daxian-dbw) [reply src/System.Management.Automation/System.Management.Automation.csproj:13]: You didn't see my reply to your comment? It's here: https://github.com/PowerShell/PowerShell/pull/15603#discussion_r989291958, and quoted below:

> It doesn't work in the following scenarios, for both VSCode and Visual Studio, which are quite common in development:
>1. Run `Start-PSBuild` first, then open `System.Management.Automation` in VSCode
> 2. Open `System.Management.Automation` in VSCode to use an existing build, like from yesterday.

File: src/System.Management.Automation/System.Management.Automation.csproj
Comment (iSazonov) [reply src/System.Management.Automation/System.Management.Automation.csproj]: I think this is fundamentally unsolvable because the code is generated on the fly and exists only at the moment of compilation. In other words OmniSharp must already be running to capture this code. 

> Also, I'm fine merging the PR without fully resolving this issue.

It's up to you. I don't mind if this workaround remains.

File: src/System.Management.Automation/System.Management.Automation.csproj
Comment (iSazonov) [reply src/System.Management.Automation/System.Management.Automation.csproj:13]: Sorry, too many comments already....

File: PowerShell.Common.props
Comment (iSazonov) [inline PowerShell.Common.props:83]: I wouldn't delete it. Might come in handy.

File: src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (iSazonov) [inline src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:105]: Maybe use `string.Create()` and interpolated string. This will make the template more readable.

File: src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.csproj
Comment (iSazonov) [inline src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.csproj:10]:
```suggestion
    <LangVersion>10.0</LangVersion>
```

File: src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.csproj
Comment (iSazonov) [inline src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.csproj:18]: It is needed only for publishing SG.
https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets
```suggestion
```

File: src/System.Management.Automation/engine/PSVersionInfo.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs:53]: Are the names follow common naming standard?
Maybe MajorVersion and so on?

File: src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs:105]: I don't think `string.Create` is available in `netstandard2.0`.
Also, IMO, it's pretty readable already.

File: src/System.Management.Automation/engine/PSVersionInfo.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/PSVersionInfo.cs:53]: What I intend to do is to make it look like `Version.Major`, `Version.Minor` and so on, to make their meaning more obvious.

As for naming standard, I did a quick search in PowerShell code base, and we do have const names using `_` in them, although most of them are const values for native code.

Personally, I think this is fine, especially that CodeFactor doesn't bark about it. But I'm willing to change if more people show concerns about the naming.

File: src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.csproj
Comment (daxian-dbw) [reply src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.csproj:18]: Good point. Updated.

File: PowerShell.Common.props
Comment (daxian-dbw) [reply PowerShell.Common.props:83]: Will bring them back.

File: src/System.Management.Automation/engine/PSVersionInfo.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/PSVersionInfo.cs:53]: It is minor. I am fine to merge as-is.

**Review by iSazonov — COMMENTED** [review]
LGTM.

File: src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (iSazonov) [reply src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs]: If we would think about third-party linux repository maintainer they have to build from sources for security reasons. And they have to learn and port Build.psm1 and other helpers again and again. They must also support previous versions (security patches). It is obvious how difficult this is. If we add that last year git tags for some reason have problems all the time, they have a big headache. It would be much easier if it just worked `dotnet publish`.

Comment (daxian-dbw) [reply src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs]: There are too many `if's` here :)
As far as I understand, it's not a goal as of today to build PowerShell completely without scripts due to the complexity of the PowerShell release build.

But, even if things were changed in future, parsing version in source generator code would not be a blocker to that goal -- it would work well with just `dotnet publish`.

File: src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (daxian-dbw) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs]: It's not clear if there is really a code injection vulnerability. The version properties are defined in `PowerShell.Common.props`, and then passed in by using the .NET source generator property `CompilerVisibleProperty`, so unless the properties are already tampered in `PowerShell.Common.props`, there seems no other way to cause a security problem, in my personal understanding.

Be noted that, today, the `ProductVersion` embedded in PowerShell assemblies is calculated using the same properties in `PowerShell.Common.props`. Then [in the static constructor](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/PSVersionInfo.cs#L77-L109) of `PSVersionInfo`, we get the `ProductVersion` from reflection, and parse the string to get PowerShell version and `GitCommitId`. So, if those properties could be tampered, then it's already a problem to our code today.

@TravisEz13 it would be very helpful if you can elaborate how the code injection could happen, and how would it affect our release builds or daily builds. Thanks!

**Comment by PaulHigin** [comment]
@WG-Security
The security group is Ok with this, because for release builds the generated version comes from a trusted source (releaseTag).

**Review by daxian-dbw — APPROVED** [review]
LGTM

File: src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs
Comment (daxian-dbw) [reply src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs]: Comment resolved by https://github.com/PowerShell/PowerShell/pull/15603#issuecomment-1273819725

**Comment by daxian-dbw** [comment]
@iSazonov Just FYI, I re-did the conflict resolution because you accidentally brough back the `PSCompatibleVersions` property, which was deleted in the other PR.

**Comment by iSazonov** [comment]
@daxian-dbw Thanks for working on the SG!.

**Comment by iSazonov** [comment]
> @iSazonov Just FYI, I re-did the conflict resolution because you accidentally brough back the `PSCompatibleVersions` property, which was deleted in the other PR.

Gihub did show only one merge conflict for
```
       internal static string ProductVersion { get; }
        internal static string GitCommitId
        {
            get
            {
                return (string)s_psVersionTable[PSGitCommitIdName];
            }
        }
```
but did not marked  PSCompatibleVersions. I wonder why. 😕

**Comment by daxian-dbw** [comment]
> Since you mentioned saving build time I think you could be interesting in SG for TypeGen.

I don't know if that would save build time, I doubt it :)

**Pull request merged**
